addons/delivery/models/sale_order.py

```diff
@@ -48,36 +45,52 @@ def delivery_set(self):
         # Remove delivery products from the sale order
         self._delivery_unset()
 
+        line_ids = []
         for order in self:
-            if order.carrier_id and order.carrier_id.delivery_type != 'grid':
-                # Shipping providers are used when delivery_type is other than 'grid'
-
-                if order.state not in ('draft', 'sent'):
-                    raise UserError(_('The order state have to be draft to add delivery lines.'))
-
-                carrier = order.carrier_id
-                account_id = carrier.product_id.property_account_income.id
-                if not account_id:
-                    account_id = carrier.product_id.categ_id.property_account_income_categ.id
-
-                # Apply fiscal position
-                taxes = carrier.product_id.taxes_id
-                taxes_ids = taxes.ids
-                if order.partner_id and order.fiscal_position_id:
-                    account_id = order.fiscal_position_id.map_account(account_id)
-                    taxes_ids = order.fiscal_position_id.map_tax(taxes).ids
-
-                SaleOrderLine.create({
-                    'order_id': order.id,
-                    'name': order.carrier_id.name,
-                    'product_uom_qty': 1,
-                    'product_uom': order.carrier_id.product_id.uom_id.id,
-                    'product_id': order.carrier_id.product_id.id,
-                    'price_unit': order.carrier_id.get_shipping_price_from_so(order)[0],
-                    'tax_id': [(6, 0, taxes_ids)],
-                    'is_delivery': True
-                })
 
+            if order.state not in ('draft', 'sent'):
+                raise UserError(_('The order state have to be draft to add delivery lines.'))
+
+            carrier = order.carrier_id
+            if carrier and carrier.delivery_type not in ['fixed', 'base_on_rule']:
+                # Shipping providers are used when delivery_type is other than 'fixed' or 'base_on_rule'
+                price_unit = order.carrier_id.get_shipping_price_from_so(order)[0]
             else:
                 # Classic grid-based carriers
-                super(SaleOrder, self).delivery_set()
+                carrier = order.carrier_id.verify_carrier(order.partner_shipping_id)
```

Comment:
hello @dmo-odoo ,
Actually carrier is the order.carrier_id some code are repeated so i take it into the carrier so condition is not removed  so not getting crash with this. see line: https://github.com/odoo/odoo/pull/7386/files#diff-2b915a4edd7a71789aaf393364824331R54 
thanks